pkg/kubelet/kubelet.go

```diff
@@ -318,6 +319,11 @@ func makePodSourceConfig(kubeCfg *kubeletconfiginternal.KubeletConfiguration, ku
 }
 
 func getRuntimeAndImageServices(remoteRuntimeEndpoint string, remoteImageEndpoint string, runtimeRequestTimeout metav1.Duration) (internalapi.RuntimeService, internalapi.ImageManagerService, error) {
+	if remoteRuntimeEndpoint == fakeremote.FakeRemoteRuntimeEndpoint && remoteImageEndpoint == fakeremote.FakeRemoteImageEndpoint {
```

Comment:
This is potentially controversial, because we're introducing dependency from fake implementation in the production kubelet code.

I think it would be cleaner to allow injecting RuntimeService and ImageManagerService explicitly (via Dependencies), and using this function only if those are unspecified.

@kubernetes/sig-node-pr-reviews - for thoughts

Reply:
It's an implement of https://github.com/kubernetes/kubernetes/issues/41504#issuecomment-412996401

Reply:
Not sure I understand - what I described above would also be implementation of that comment.

Reply:
> allow injecting RuntimeService and ImageManagerService explicitly (via Dependencies), and using this function only if those are unspecified.

Not completely understand it. Seem it will refactor more kubelet code?

Reply:
Sorry - I meant that this change means that kubelet.go file starts depending on fake implementation of CRI, which is purely a test code (so prod code depends on test code, which is always a bad situation).

So what I'm suggesting is that kubeletDeps may potentially support specifying those things, and `getRuntimeAndImageServices` is called only if those aren't specified.
So that will allow you to not change this function (and introduce dependency on the fake implementation), and only do that in hollowNode explicitly (which is test code anyway).

Reply:
> prod code depends on test code, which is always a bad situation.

completely agree
> So what I'm suggesting is that kubeletDeps may potentially support specifying those things, and `getRuntimeAndImageServices` is called only if those aren't specified.
> So that will allow you to not change this function (and introduce dependency on the fake implementation), and only do that in hollowNode explicitly (which is test code anyway).

get, thank you.

Reply:
> I think it would be cleaner to allow injecting RuntimeService and ImageManagerService explicitly (via Dependencies), and using this function only if those are unspecified.

Done, clean some kubelet code at the same time.

Reply:
Thanks - this looks good to me.
But I would like to hear also from Kubelet owners before approving.